### PR TITLE
Payment mutations

### PR DESCRIPTION
**Pull request by korycins**
This PR adds two new mutations that can be used to handle payments assigned to order/checkout by external services like CheckoutApp.  It is a second API part related to the PR https://github.com/saleor/saleor/pull/9343.  There are some FIXME comments left but they will be handled in separate PR with handling Payment models. This introduces only new API changes.
Below charts should give more context how we want to use a new mutations.

### Case: Order before payment
```mermaid
sequenceDiagram
    Customer->>+Checkout App: Finalize an order
    Checkout App ->>+ Saleor: Create order from checkout
    Saleor -->>- Checkout App: Unpaid order
    Checkout App-->>-Customer: Order
    Customer ->>+ Checkout App: Process payment
    Checkout App ->>+ Payment provider: Process payment for order
    Payment provider -->>- Checkout App: Payment with proper status
    rect rgb(162, 162, 162)
    Note over Checkout App, Saleor: Covered part by this PR
    Checkout App ->>+ Saleor: Create payment for order
    Saleor -->>- Checkout App: Payment created and <br>attached to order
    end
    Checkout App -->>- Customer: Order paid
```

### Case:  Payment before order
```mermaid
sequenceDiagram
    Customer ->>+ Checkout App: Process payment
    Checkout App ->>+ Payment provider: Process payment for checkout
    Payment provider -->>- Checkout App: Payment with proper status
    rect rgb(162, 162, 162)
    Note over Checkout App, Saleor: Covered part by this PR
    Checkout App ->>+ Saleor: Create payment for checkout
    Saleor -->>- Checkout App: Payment created and <br>attached to checkout
    end
    Checkout App -->>- Customer: Payment paid
    Customer->>+Checkout App: Conver checkout into an order
    Checkout App ->>+ Saleor: Create order from checkout
    Saleor ->> Saleor: Attach checkout's payments to order
    Saleor -->>- Checkout App: Paid order
    Checkout App-->>-Customer: Order
```

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/order/types.py [hunk 1]
```diff
@@ -1015,9 +1106,16 @@ def _resolve_payment_status(payments):
     @staticmethod
     def resolve_payment_status_display(root: models.Order, info):
         def _resolve_payment_status(payments):
-            if last_payment := get_last_payment(payments):
+            last_payment = get_last_payment(payments)
+            # FIXME these conditions will be changed in separate PR after converting
+            #  current payment model into LegacyPayment object
+            if not last_payment:
+                return dict(ChargeStatus.CHOICES).get(ChargeStatus.NOT_CHARGED)
```
Comment (L3str4nge) [inline saleor/graphql/order/types.py:1113]: Wouldn't be better to return this choice inline?

File: saleor/graphql/payment/types.py [hunk 2]
```diff
@@ -81,42 +81,82 @@ class Meta:
...
+        deprecation_reason=f"{DEPRECATED_IN_3X_FIELD} Use new checkout flow."
+    )
+    charge_status = PaymentChargeStatusEnum(
+        description="Internal payment status.",
+        required=True,
+        deprecation_reason=f"{DEPRECATED_IN_3X_FIELD} Use new checkout flow.",
     )
     transactions = graphene.List(
-        Transaction, description="List of all transactions within this payment."
+        Transaction,
```
Comment (fowczarek) [inline saleor/graphql/payment/types.py:136]:
```suggestion
        graphene.NonNull(Transaction),
```

**Review by fowczarek — APPROVED** [review]

**Review by L3str4nge — APPROVED** [review]

File: saleor/graphql/payment/types.py (hunk 2 above)
Comment (korycins) [reply saleor/graphql/payment/types.py:136]: It is covered in separate PR:
https://github.com/saleor/saleor/pull/9391/files#diff-a5ec8861a37bb971cbfe0b5fff62d4c509e018f976e5735b6f2a37ee0f11bfc1R107

File: saleor/graphql/order/types.py [hunk 3]
```diff
@@ -125,6 +127,25 @@ def get_order_discount_event(discount_obj: dict):
...
+    status = ChargeStatus.NOT_CHARGED
+    captured_money = prices.Money(Decimal(0), order.currency)
+    refunded_money = prices.Money(Decimal(0), order.currency)
+    for payment in payments:
+        captured_money += payment.amount_captured
+        refunded_money += payment.amount_refunded
+
+    if captured_money >= order.total.gross:
+        status = ChargeStatus.FULLY_CHARGED
+    if captured_money and captured_money <= order.total.gross:
```
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/graphql/order/types.py:140]: > Also captured_money < order.total.gross instead, as the previous condition already cheks the equality.



I think that `captured_money and captured_money <= order.total.gross:`  needs to stay, as I don't want to set status `PARTIALLY_CHARGED` when `captured_money` == 0

File: saleor/graphql/order/types.py (hunk 1 above)
Comment (korycins) [reply saleor/graphql/order/types.py:1113]: `ChargeStatus.CHOICES` is a list of (str, str). I need to map it to dict, to get a readable description of status. Not sure if I can change it. I could only replace it with `str` but not sure if this woul be a best way for that

**Pull request merged**